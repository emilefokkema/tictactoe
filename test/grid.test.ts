import { describe, it, expect } from 'vitest'
import { Grid } from '../src/scripts/content/grid/grid'
import { GridCellMeasurements } from '../src/scripts/content/grid/types'
import { Theme } from '../src/scripts/themes'
import { mockContext } from './mock-context'

describe('a grid', () => {
    const measurements: GridCellMeasurements = {
        x: 0,
        y: 0,
        size: 100,
        background: {
            extendLeft: 0,
            extendRight: 0,
            extendTop: 0,
            extendBottom: 0
        }
    }
    const lightTheme: Theme = {
        color: '#000',
        backgroundColor: '#fff',
        get loserTheme(){return darkTheme;},
        get winnerTheme(){return lightTheme;}
    };
    const darkTheme: Theme = {
        color: '#fff',
        backgroundColor: '#000',
        get loserTheme(){return darkTheme;},
        get winnerTheme(){return lightTheme;}
    }


    it('should draw borders', () => {
        const context = mockContext();
        const grid = new Grid(measurements, lightTheme);
        grid.drawBorders(context.ctx);

        expect(context.record).toMatchSnapshot();
    })

    it('should change themes', () => {
        const context = mockContext();
        const grid = new Grid(measurements, lightTheme);
        grid.setTheme(darkTheme);

        grid.drawBorders(context.ctx);

        expect(context.record).toMatchSnapshot();

        context.reset();

        grid.cells[0].setTheme(lightTheme)
        grid.drawCells(context.ctx);

        expect(context.record).toMatchSnapshot();
    });

    it('should theme two adjacent cells', () => {
        const context = mockContext();
        const grid = new Grid(measurements, lightTheme);
        grid.cells[0].setTheme(darkTheme);
        grid.cells[1].setTheme(darkTheme);

        grid.drawBorders(context.ctx);

        expect(context.record).toMatchSnapshot();

        context.reset();

        grid.setTheme(darkTheme);

        grid.drawBorders(context.ctx);

        expect(context.record).toMatchSnapshot();
    })

    it('should theme six adjacent cells', () => {
        const context = mockContext();
        const grid = new Grid(measurements, lightTheme);
        grid.cells[0].setTheme(darkTheme);
        grid.cells[1].setTheme(darkTheme);
        grid.cells[2].setTheme(darkTheme);
        grid.cells[3].setTheme(darkTheme);
        grid.cells[4].setTheme(darkTheme);
        grid.cells[5].setTheme(darkTheme);

        grid.drawBorders(context.ctx);

        expect(context.record).toMatchSnapshot();
    })
})