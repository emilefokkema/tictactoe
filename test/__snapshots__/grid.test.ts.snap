// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`a grid > should change themes 1`] = `
[
  "ctx.save();",
  "ctx.lineWidth = 1",
  "ctx.strokeStyle = "#fff"",
  "ctx.beginPath();",
  "ctx.moveTo(33.166666666666664, 0);",
  "ctx.lineTo(33.166666666666664, 100);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(66.83333333333334, 0);",
  "ctx.lineTo(66.83333333333334, 100);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 33.166666666666664);",
  "ctx.lineTo(100, 33.166666666666664);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 66.83333333333334);",
  "ctx.lineTo(100, 66.83333333333334);",
  "ctx.stroke();",
  "ctx.restore();",
]
`;

exports[`a grid > should change themes 2`] = `
[
  "ctx.fillStyle = "#fff"",
  "ctx.fillRect(0, 0, 33.166666666666664, 33.166666666666664);",
]
`;

exports[`a grid > should draw borders 1`] = `
[
  "ctx.save();",
  "ctx.lineWidth = 1",
  "ctx.strokeStyle = "#000"",
  "ctx.beginPath();",
  "ctx.moveTo(33.166666666666664, 0);",
  "ctx.lineTo(33.166666666666664, 100);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(66.83333333333334, 0);",
  "ctx.lineTo(66.83333333333334, 100);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 33.166666666666664);",
  "ctx.lineTo(100, 33.166666666666664);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 66.83333333333334);",
  "ctx.lineTo(100, 66.83333333333334);",
  "ctx.stroke();",
  "ctx.restore();",
]
`;

exports[`a grid > should theme six adjacent cells 1`] = `
[
  "ctx.save();",
  "ctx.lineWidth = 1",
  "ctx.strokeStyle = "#000"",
  "ctx.save();",
  "ctx.strokeStyle = "#fff"",
  "ctx.beginPath();",
  "ctx.moveTo(33.166666666666664, 0);",
  "ctx.lineTo(33.166666666666664, 67.33333333333334);",
  "ctx.stroke();",
  "ctx.restore();",
  "ctx.beginPath();",
  "ctx.moveTo(33.166666666666664, 67.33333333333334);",
  "ctx.lineTo(33.166666666666664, 100);",
  "ctx.stroke();",
  "ctx.save();",
  "ctx.strokeStyle = "#fff"",
  "ctx.beginPath();",
  "ctx.moveTo(66.83333333333334, 0);",
  "ctx.lineTo(66.83333333333334, 67.33333333333334);",
  "ctx.stroke();",
  "ctx.restore();",
  "ctx.beginPath();",
  "ctx.moveTo(66.83333333333334, 67.33333333333334);",
  "ctx.lineTo(66.83333333333334, 100);",
  "ctx.stroke();",
  "ctx.save();",
  "ctx.strokeStyle = "#fff"",
  "ctx.beginPath();",
  "ctx.moveTo(0, 33.166666666666664);",
  "ctx.lineTo(100, 33.166666666666664);",
  "ctx.stroke();",
  "ctx.restore();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 66.83333333333334);",
  "ctx.lineTo(100, 66.83333333333334);",
  "ctx.stroke();",
  "ctx.restore();",
]
`;

exports[`a grid > should theme two adjacent cells 1`] = `
[
  "ctx.save();",
  "ctx.lineWidth = 1",
  "ctx.strokeStyle = "#000"",
  "ctx.save();",
  "ctx.strokeStyle = "#fff"",
  "ctx.beginPath();",
  "ctx.moveTo(33.166666666666664, 0);",
  "ctx.lineTo(33.166666666666664, 32.666666666666664);",
  "ctx.stroke();",
  "ctx.restore();",
  "ctx.beginPath();",
  "ctx.moveTo(33.166666666666664, 32.666666666666664);",
  "ctx.lineTo(33.166666666666664, 100);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(66.83333333333334, 0);",
  "ctx.lineTo(66.83333333333334, 100);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 33.166666666666664);",
  "ctx.lineTo(100, 33.166666666666664);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 66.83333333333334);",
  "ctx.lineTo(100, 66.83333333333334);",
  "ctx.stroke();",
  "ctx.restore();",
]
`;

exports[`a grid > should theme two adjacent cells 2`] = `
[
  "ctx.save();",
  "ctx.lineWidth = 1",
  "ctx.strokeStyle = "#fff"",
  "ctx.beginPath();",
  "ctx.moveTo(33.166666666666664, 0);",
  "ctx.lineTo(33.166666666666664, 100);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(66.83333333333334, 0);",
  "ctx.lineTo(66.83333333333334, 100);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 33.166666666666664);",
  "ctx.lineTo(100, 33.166666666666664);",
  "ctx.stroke();",
  "ctx.beginPath();",
  "ctx.moveTo(0, 66.83333333333334);",
  "ctx.lineTo(100, 66.83333333333334);",
  "ctx.stroke();",
  "ctx.restore();",
]
`;
